nixos/knot: add keyFiles option

Mic92 opened the pull request:
Add an option to include configuration not in the nix store.
The use case for this are nixops like tools where the configuration
is not world-readable. Because of knot's DynamicUser we need to
install those files with the correct permission on startup into /var/lib/knot.

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Mic92 commented:
Because of `DynamicUser` the permission management is not yet very straight forward:

```  
services.knot = {
  enable = true;
  checkConfig = false;
  extraConfig = ''
    include: /var/lib/knot/knot-he-key.conf
  '';
};
systemd.services.knot = {
  serviceConfig.PermissionsStartOnly = true;
  preStart = ''
    install -m700 --owner $USER /var/src/secrets/knot-he-key.conf /var/lib/knot/knot-he-key.conf
  '';
};
```

Maybe rather than `checkConfig` I could also add an option to add custom secret files?

mweinelt reviewed (approved):
I understand your use case and the change is fine with me.

vcunat commented:
I haven't really thought about best practices of managing these secrets (for knot service).

Mic92 commented:
I though about adding an `systemd.services.knot.keyFiles = [ "tsig.conf" ];`.
Those would be copied to `/var/lib/knot/keys.d` and than included in knot's default configuration.

vcunat commented:
I wonder if the risk of omitting the quotation marks is too high :thinking:

Mic92 commented:
> I wonder if the risk of omitting the quotation marks is too high thinking

Usually when using nixops is using something like `/run/keys`, which is likely to fail on the developer machine. In theory all passwordFile options in nixos could be accidentally miss-used like that. However I see a fix for that outside of scope of this PR. This would be solved by a linter or custom nixos type option that check if the path is in the nix store.

symphorien commented:
About passwordFile like options: https://github.com/NixOS/nixpkgs/pull/78640

Mic92 committed:
knot: put runtime paths outside the nix store

Otherwise knot tries to write to non-writable directories.
This for example breaks dnssec signing.
While it's possible to overwrite these path in the configuration,
having a sane defaults is nicer.

Mic92 commented:
With the latest commit knot now can do automatic DNSSEC signing.

Mic92 commented on nixos/modules/services/networking/knot.nix:
```diff
@@ -5,14 +5,16 @@ with lib;
 
-  configFile = pkgs.writeText "knot.conf" cfg.extraConfig;
-  socketFile = "/run/knot/knot.sock";
+  configFile = pkgs.writeTextFile {
```
I tried to convert nix attrsets to yaml using `remarshal`. However the yaml subset chosen by knot is too restrictive for that (i.e. expect keys to appear in a certain order).

vcunat replied:
Have you seen [this thread](https://github.com/NixOS/nixpkgs/pull/56922#issuecomment-470740544)?  If a few people tries this out in practice, I suppose we could make it available as another optional attribute.

vcunat replied:
I'm not aware of keys being expected in particular order, though (perhaps I misunderstand).

Mic92 replied:
I had an issue with an acl rule where the `id` key was not the first key, which lead to an error. This broke already in the nixos test. Anyway this PR should not interfere with migrating to a different contribute type...

vcunat replied:
Oh, that's a known restriction, though not by me until now.

vcunat replied:
I think I have an idea how to approach this; hopefully I can get it to work soon.  I'll link from here when that happens.

vcunat replied:
It didn't really go the way I imagined, but let's continue on https://github.com/NixOS/nixpkgs/pull/81460

Mic92 commented:
Instead of adding an option to disable configuration checks I added a `keyFiles` option that implicitly disable configuration checks when used. This also gets permissions right in combination with knot's DynamicUser option.

Mic92 commented:
Here is a real-world example: https://github.com/Mic92/dotfiles/blob/34a97d3292b833fbad8f716f9c37a28eb4121eb1/nixos/eve/modules/knot/default.nix

Also the nixos test covers the new feature as well.

Mic92 committed:
knot: drop dynamic user

This makes it hard to include secret files.
Also using tools like keymgr becomes harder.

Mic92 committed:
knot: add keyFiles option

This useful to include tsig keys using nixops without adding those
world-readable to the nix store.

Mic92 commented:
I dropped DynamicUser. Also because of https://github.com/NixOS/nixpkgs/pull/79928

Mic92 commented:
@ju1m If you would have a short look at it and give me a thumbs up, I would go forward and merge it!
